File: min-camlMy/elimEqExp.ml
(* 共通部分式除去 *)
(* assoc(AlphaNormal)の後が良い?? *)
(*
副作用のある外部関数(アセンブリで書いたライブラリ)は初期集合にいれとく
*)

open KNormal

module Mexp =
  Map.Make
    (struct
      type t = KNormal.t
      let compare = Pervasives.compare
     end)

(* effect :: KNormal.t -> Bool *)
let rec effect set = function (* 副作用の条件とは?? *)
  | Let(_,e1,e2)| IfEq(_,_,e1,e2)| IfLE(_,_,e1,e2) -> effect set e1 || effect set e2
  | LetRec(_,e) | LetTuple(_,_,e) -> effect set e
  | App (f, xs)	when S.mem f set	(* fは関数名(再帰関数だったらfalse) *)
      -> ((*Format.eprintf "side effect fun(elimEqExp) %s@." f;*)
	  true)
  | ExtFunApp (f, xs) when S.mem f set	(* fは関数名 *)
      -> ((*Format.eprintf "side effect fun(elimEqExp) %s@." f;*)
	  true)
  | Put _ | Get _ -> true
  | Var x when S.mem x set
      -> ((*Format.eprintf "side effect var(elimEqExp) %s@." x;*)
	  true)
  | _ -> false

(* env :: map(KNormal.t * Id.t), set :: set(Id.t) side effect function *)
let rec g env set = function
  | IfEq(x, y, e1, e2) -> IfEq(x, y, g env set e1, g env set e2)
  | IfLE(x, y, e1, e2) -> IfLE(x, y, g env set e1, g env set e2)
  | Let((x, t), e1, e2) ->
    let e1' = g env set e1 in
    if effect set e1' 	(*副作用の有無*)
    then ((*Format.eprintf "side effect var(elimEqExp) %s@." x;*)
	  Let((x, t), e1', g env (S.add x set) e2))
    else 
      if Mexp.mem e1' env
      then (* mapにあったら変換。登録しない?*)
	((match e1' with
	  | App(fname,_) -> Format.eprintf "elimEqExp(FUN) %s@." fname
	  | _  -> ());
	 let y = Mexp.find e1' env in
	 Format.eprintf "elimEqExp %s = %s@." x y;
	 Let((x, t), Var(y), g env (S.add x set) e2))
      else Let((x, t), e1', g (Mexp.add e1' x env) (S.add x set) e2)(* 登録する *)
  | LetRec({ name = (x, t); args = yts; body = e1 }, e2) ->
    if effect set e1	(* 再帰呼び出しだったらどうする?? *)
    then
      ((*Format.eprintf "side effect fun(elimEqExp) %s@." x;*)
       LetRec({ name = (x, t); args = yts; body = g (Mexp.empty) set e1 }, g env (S.add x set) e2))
    else
      LetRec({ name = (x, t); args = yts; body = g (Mexp.empty) set e1 }, g env set e2)
  | LetTuple(xts, y, e) -> LetTuple(xts, y, g env set e)
  | e -> e

let f = g Mexp.empty (S.of_list ["create_array"; "create_float_array"; "print_char"; "input_char"])

File: hs-mincaml/ElimEqExp.hs
module ElimEqExp where

import qualified IdMod as I
import qualified KNormal as K

import qualified Data.Set as St
import qualified Data.Map as Mp
import Control.Monad.State
import qualified Debug.Trace as DT

type EEnv = Mp.Map K.T I.Id
type EffectSet = St.Set I.Id    -- side effect variables

type EqState = State EffectSet 

initSet = St.fromList [ "create_array"  
                      , "create_float_array"
                      , "print_char"
                      , "input_char" ]
  
elimEqExpMain :: K.T -> K.T
elimEqExpMain exp = -- let !_ = DT.trace (show eSet) () in
  f Mp.empty exp
  where (_, eSet) = runState (effect exp) initSet
        
        f :: EEnv -> K.T -> K.T
        f eEnv exp = case exp of
          K.Let (x,t) e1 e2 ->
            if St.member x eSet 
            then K.Let (x,t) e1' e2'
            else (if Mp.member e1' eEnv
                  then let !_ = DT.trace ("ElimEqExp: "++(show x)++" = "++(show y)) () in
                       K.Let (x,t) (K.Var y) e2'
                  else K.Let (x,t) e1' e2'')
              where e1'    = f eEnv e1
                    e2'    = f eEnv e2                
                    Just y = Mp.lookup e1' eEnv
                    e2''   = f (Mp.insert e1' x eEnv) e2
          K.LetRec K.Fundef{K.name=(x,t), K.args=yts, K.body=e1} e2 -> 
            K.LetRec K.Fundef{K.name=(x,t), K.args=yts, K.body=(f Mp.empty e1)} (f eEnv e2)
          K.LetTuple xts y e    -> K.LetTuple xts y (f eEnv e)
          K.IfEq x y e1 e2      -> K.IfEq x y (f eEnv e1) (f eEnv e2)
          K.IfLe x y e1 e2      -> K.IfLe x y (f eEnv e1) (f eEnv e2)
          _                     -> exp
          
-- alphaTrans必須
effect :: K.T -> EqState Bool
effect exp = case exp of
  K.Let (x,t) e1 e2     -> 
    do b1 <- effect e1
       if b1 then do modify (\set-> St.insert x set)
                     b2 <- effect e2
                     return (b1 || b2)
             else do b2 <- effect e2
                     return (b1 || b2)                        
  K.LetRec K.Fundef{K.name=(x,t), K.args=yts, K.body=e1} e2 ->
    do b1 <- effect e1
       if b1 then do modify (\set-> St.insert x set)
                     effect e2
             else effect e2
  K.LetTuple xts y e    -> 
    do b1 <- gets (St.member y) 
       if b1 then do modify (\set-> foldl (\set (x,t)-> St.insert x set) set xts)
                     effect e
             else effect e
  K.App x ys            -> gets (\set -> or $ map (\y -> St.member y set) (x:ys))
  K.ExtFunApp x ys      -> gets (\set -> or $ map (\y -> St.member y set) (x:ys))
  K.IfEq x y e1 e2      -> do [b1,b2] <- mapM effect [e1,e2]; return (b1 || b2)
  K.IfLe x y e1 e2      -> do [b1,b2] <- mapM effect [e1,e2]; return (b1 || b2)
  K.Get _ _             -> return True
  K.Put _ _ _           -> return True
  _                     -> return False
  